cmd/log/src/lib.rs

```diff
@@ -0,0 +1,276 @@
+    variable: &'a HubrisVariable,
+) -> Result<&'a str> {
+    Ok(&hubris.lookup_module(HubrisTask::from(variable.goff))?.name)
+}
+
+// this allow is meant for the header println! in the body but you cannot apply
+// an attribute to a macro invoction, so we have to put it here instead.
+#[allow(clippy::print_literal)]
+fn stringbuf(context: &mut humility::ExecutionContext) -> Result<()> {
+    let hubris = context.archive.as_ref().unwrap();
+    let core = &mut **context.core.as_mut().unwrap();
+    let Subcommand::Other(subargs) = context.cli.cmd.as_ref().unwrap();
+    let subargs = StringbufArgs::try_parse_from(subargs)?;
+
+    let mut ringbufs = vec![];
+
+    for v in hubris.qualified_variables() {
+        if let Some(ref name) = subargs.name {
+            if v.0.eq(name)
+                || (v.0.ends_with("_STRINGBUF")
+                    && (v.0.contains(name)
+                        || taskname(hubris, v.1)?.contains(name)))
+            {
+                ringbufs.push(v);
+            }
+        } else if v.0.ends_with("_STRINGBUF") {
+            ringbufs.push(v);
+        }
+    }
+
+    if ringbufs.is_empty() {
+        if let Some(name) = subargs.name {
+            bail!("no ring buffer name contains \"{}\" (-l to list)", name);
+        } else {
+            bail!("no ring buffers found");
+        }
+    }
+
+    if subargs.monitor && ringbufs.len() != 1 {
+        if let Some(name) = subargs.name {
+            bail!(
+                "\"{}\" matched more than one stringbuf (-l to list all)",
+                name
+            );
+        } else {
+            bail!("found more than one stringbuf, please specify a name");
+        }
+    }
+
+    ringbufs.sort();
+
+    if subargs.list {
+        println!("{:18} {:<30} {:<10} {}", "MODULE", "BUFFER", "ADDR", "SIZE");
+
+        for v in ringbufs {
+            let t = taskname(hubris, v.1)?;
+
+            println!("{:18} {:<30} 0x{:08x} {:<}", t, v.0, v.1.addr, v.1.size);
+        }
+
+        return Ok(());
+    }
+
+    for v in ringbufs {
+        // Try not to use `?` here, because it causes one bad ringbuf to make
+        // them all unavailable.
+        println!(
+            "humility: stringbuf {} in {}:",
+            v.0,
+            taskname(hubris, v.1).unwrap_or("???")
+        );
+        if let Ok(def) = hubris.lookup_struct(v.1.goff) {
+            if subargs.monitor {
+                if let Err(e) =
+                    stringbuf_monitor(hubris, core, def, v.1, subargs.delay)
+                {
+                    humility::msg!("stringbuf monitor cancelled: {}", e);
+                }
+            } else {
+                // this first read is just to get the last entry, should probably change to reading
+                // just stringbuf.last...
+                let (_log, last) = stringbuf_read(hubris, core, def, v.1, 0)?;
+                // now reuse that last to ensure we read the whole buffer in the correct order
```

Comment:
I am confused. Why do you need to call `stringbuf_read` twice?

Reply:
The first read is to read the value of `stringbuf.last`.  Here I was using `stringbuf_read` to get the value, but that was just a temporary hack while I refactored out the simpler function `load_stringbuf`.

The reason we need last is that `stringbuf_read` uses it to determine where new characters in the buffer are, and print them out in order.  Without the correct value of last, `stringbuf_read` could print the log characters starting at the beginning of the buffer, not from the oldest character.  It would technically be possible to do this without having to read the stringbuf twice, but it would require writing a new variation on `stringbuf_read`, and I don't think that it is worth the effort.

Reply:
I see what you are saying. I think we should simplify the logic in stringbuf_read. Anyways, we have gone back and forth a lot of times here, so we can push this in and follow up with some clean up changes.

Reply:
I went ahead and simplified things.